### PR TITLE
 Set `autovacuum=on` on coordinator as default to enable auto-ANALYZE.

### PR DESCRIPTION
**Pull request by JunfengYang**
Enable auto-ANALYZE as default for the cluster through set GUC's default
value. And make all test cases pass. To summarise, for all of the GPDB test
cases, there are 3 classes of cases that need to deal with autovacuum
auto-ANALYZE enabled.

1. ADD `ANALYZE` for tables.
For this kind of case, we used to have `autostats` enabled, and our
default `gp_autostats_mode` setting is `on_no_stats`.
This means `ANALYZE` is executed when first-time `insert/copy` data into
the **empty** table. So our plan is base on this.
When we enable auto-ANALYZE, `autostats` is skipped. So the table will
not have any stats until the auto-ANALYZE triggered or explicit
`ANALYZE` issued.
The failure could be caused by the explain output. The planner or ORCA
will generate a plan base on empty stats.
`ANALYZE` the tables after the first `insert/copy` could resolve the
issue, and it'll not change the plan.
But there still some cases that the test contains multiple `insert/copy`
statements and the auto-ANALYZE may run after `insert/copy` statements.
Which makes the test flaky since the plan may change. So I have to add
the `ANALYZE` after all `insert/copy` statements and modify the plan in
the answer file.
But for this case, we should be careful, some of the tests are expected
to have a specified plan(e.g. hash join), this will need more consideration.

2. Disable `autovacuum` for specific tests.
There are some cases we need to turn off the `autovacuum`.
    - Test cases related to `autostats`. These tests are testing the
    `autostats`, so we have to disable `autovacuum`.
    - Test cases are sensitive to backend transactions. The `vacuum`
    related tests. The auto-ANALYZE worker keeps the recently dead
    tuples and ao segfiles.

3. Rewrite queries to have consistent output. Which also cause
by the backend transactions from auto-ANALYZE workers.
Some of the tests query the segfile status. For example:
    ```
    VACUUM ao;
    SELECT segno, tupcount FROM gp_ao_or_aocs_seg('ao')
    ```
    If a backend transaction is running before the VACUUM, the
    output of the query will show segfile with state 2(AOSEG_STATE_AWAITING_DROP).

NOTE:
When adding new cases, please consider whether the case will be affected
by the autovacuum worker process, it may become flaky.

The `gp_autostats_mode` is set to off by default. But users can enable it manually.
The `auto_stats` and auto-ANALYZE features are separate logic and can be enabled
at the same time since these two features work at different granularities.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
I would like the PR to be split in exactly those 3 separate commits as mentioned in description. It becomes easy to know which test is modified for what purpose.

> When we enable auto-ANALYZE, autostats is skipped.

Also, missing why this is the case? What's the problem with keeping `gp_autostats_mode` to `on_no_stats` till we receive the feedback.

File: gpMgmt/bin/gpinitsystem [hunk 1]
```diff
@@ -1209,6 +1210,10 @@ CREATE_QD_DB () {
 		ERROR_CHK $? "set log_statement=all in ${GP_DIR}/$PG_CONF" 1
 		LOG_MSG "[INFO]:-Setting Master instance check point segments"
 
+		LOG_MSG "[INFO]:-Enable auto-ANALYZE for the cluster"
```
Comment (ashwinstar) [inline gpMgmt/bin/gpinitsystem:1213]: Why are we doing this via gpinitsystem, instead of just changing the default value of the GUC in code?

File: src/backend/catalog/namespace.c [hunk 2]
```diff
@@ -3302,6 +3302,10 @@ isTempNamespaceInUse(Oid namespaceId)
  * namespace (either my own, or another backend's), return the BackendId
  * that owns it.  Temporary-toast-table namespaces are included, too.
  * If it isn't a temp namespace, return InvalidBackendId.
+ *
+ * FIXME: this only works if the temp namespace is created in utility mode.
```
Comment (ashwinstar) [inline src/backend/catalog/namespace.c:3306]: It's not clear what this FIXME is trying to convey, mostly from perspective of what's broken and what next action is expected from it?

File: src/backend/postmaster/README.auto-ANALYZE [hunk 3]
```diff
@@ -1,50 +1,42 @@
-Enable auto-ANALYZE through autovacuum daemon on Master(QD) in GPDB.
+Enable auto-ANALYZE through autovacuum daemon on Coordinator(QD) in GPDB.
```
Comment (ashwinstar) [inline src/backend/postmaster/README.auto-ANALYZE:1]: Thanks a lot for making the Master -> Corodinator change :+1:

**Review by ashwinstar — COMMENTED** [review]
I am missing to understand 3rd category of test failures. Please can you point to example test failure for that. Reading it seems not related to enabling auto-analyze but generic test race condition irrespective.

**Comment by ashwinstar** [comment]
> Disable autovacuum for specific tests.

For these category of tests instead of globally flipping autovacuum GUC, would it be better to create tables in those tests with `autovacuum_enabled = false` instead and keep it simple and faster.

File: gpMgmt/bin/gpinitsystem (hunk 1 above)
Comment (JunfengYang) [reply gpMgmt/bin/gpinitsystem:1213]: As mentioned in `src/backend/postmaster/README.auto-ANALYZE`:
```
### Eable auto-ANALYZE as default for GPDB cluster.
Since set the `autovacuum=on` in guc.c will take effect for all segments.
And we only need to set it on Coordinator. So, set the `autovacuum=on` for Coordinator
in gpinitsystem.
```

**Comment by JunfengYang** [comment]
> > Disable autovacuum for specific tests.
> 
> For these category of tests instead of globally flipping autovacuum GUC, would it be better to create tables in those tests with `autovacuum_enabled = false` instead and keep it simple and faster.

@ashwinstar For case 2, we could set `autovacuum_enabled = false`. But for case 1, if `autovacuum=on`, autostats will be skipped too. The logic is, we don't need auto-ANALYZE and autostats at the same time since they are doing the same thing. So if `autovacuum=on`, autostats is disabled. And using the auto-ANALYZE logic to execute ANALYZE. Otherwise, things get confusing.
Since I just group the tests which need to turn off autovacuum, so this does not generate more diffs.
But if you think `autovacuum_enabled = false` is better, I'll fix it.

File: src/backend/catalog/namespace.c (hunk 2 above)
Comment (JunfengYang) [reply src/backend/catalog/namespace.c:3306]: How about
```
* FIXME: this only works if the temp namespace is created in utility mode.
* Since the temp namespace for QD and QE is using gp_session_id as suffix
* instead of backendID.
* The returned value is actually gp_session_id instead of the expected backendID
* unless the temp namespace is created in utility mode.
* We need to find a way to get the backendID for the temp namespace created
* on QD and QE.
```

File: gpMgmt/bin/gpinitsystem (hunk 1 above)
Comment (ashwinstar) [reply gpMgmt/bin/gpinitsystem:1213]: I don't like this solution. Curious what happens if autovacuum is enabled on segments, will it crash or perform something wrong or silently locally analyze tables or something else?

I feel if we only use autovacuum as *GUC* for Coordinator (ON or OFF) and for segments need to have it OFF for sure then should not treat this as GUC for segments and in code also disable the functionality for segments. Utility mode can decide to keep the functionality enabled or disabled (I think disabled is fine).

Just FYI...in near future we will be enabling auto-vacuum process to not just do auto-ANALYZE but also auto-VACUUM where we would like to have this GUC ON for master and segments.

**Comment by JunfengYang** [comment]
> Also, missing why this is the case? What's the problem with keeping gp_autostats_mode to on_no_stats till we receive the feedback.

@ashwinstar Think below case.
A cluster with `autovacuum=on`, and create a table with `autovacuum_enabled = false`.
As I user, I would expect the table will always have empty stats unless I manually run ANALYZE.
But if we still run autostats with `autovacuum=on`. The table will have stats.
So we should not run autostats with `autovacuum=on`.

>I would like the PR to be split in exactly those 3 separate commits as mentioned in description. 

Actually, all test SQL and expected output changed are belong to case 1 and 3. Case 2 is changed through schedule file.
I'll try to separate commits.

File: gpMgmt/bin/gpinitsystem (hunk 1 above)
Comment (JunfengYang) [reply gpMgmt/bin/gpinitsystem:1214]: If `autovacuum` is enabled on segments, it'll start an autovacuum launcher, but it'll only run vacuum for template0 for xact id wrap around.
So set only for Coordinator will save the autovacuum launcher process on each segments since template0 xact id wrap around will be forced. I don't see any reason why we can not set autovacuum on segments, it just we don't set as default.

As discussed at the beginning in gpdb-dev and implementation in PR: https://github.com/greenplum-db/gpdb/pull/10515. The auto-ANALYZE always triggered on Coordinator.
And for auto-vacuum, as we discussed before, there will be two solutions.
1. Same with auto-ANALYZE, triggered by Coordinator for the whole cluster.
2. Each segment mind it's own vacuum.
In either way, the current implementation will not affect.

For the first solution, set `autovacuum=on` on Coordinator will do both auto-vacuum and auto-analyze for the whole cluster. And if set `autovacuum=on` on segments, same with current behavior.

For the second solution, set `autovacuum=on` on Coordinator will do auto-ANALYZE for the whole cluster and auto-vacuum for the Coordinator. And if set `autovacuum=on` on segments, do the auto-vacuum for segment itself.

Comment (JunfengYang) [reply gpMgmt/bin/gpinitsystem:1214]: Enable `autovacuum=on` on segments will start an autovacuum launcher process. And the launcher only executes template0 xact id wrap around vacuum currently. Actually, the xact id wrap around will always be triggered with or without `autovacuum=on` on segments. But I also don't see any reason to disable the guc on segments.

As discussed at the beginning in gpdb-dev and implementation in PR: https://github.com/greenplum-db/gpdb/pull/10515. The auto-ANALYZE always triggered on Coordinator.
And for auto-vacuum, as we discussed before, there will be two solutions.
1. Same with auto-ANALYZE, triggered by Coordinator for the whole cluster.
2. Each segment mind it's own vacuum.
In either way, the current implementation will not affect.

For the first solution, set `autovacuum=on` on Coordinator will do both auto-vacuum and auto-analyze for the whole cluster. And if set `autovacuum=on` on segments, same with current behavior.

For the second solution, set `autovacuum=on` on Coordinator will do auto-ANALYZE for the whole cluster and auto-vacuum for the Coordinator. And if set `autovacuum=on` on segments, do the auto-vacuum for segment itself.

If it's still not clear for my explain, maybe we can find a time to pair on this?

**Review by ashwinstar — COMMENTED** [review]
I think we should explicitly discuss on gpdb-users mailing list if turning off gp_autostats_mode to none is good idea or not. I think the impact to users seems similar to what you had to do via commit "Add ANALYZE for tables to pass tests." Need to find all the places in code where they need to go and explicitly add manual ANALYZE to retain existing behaviour. I understand the background worker will kick in and analyze the table but that happens async way compared to current sync way. And behaviour difference is something I am concerned with. I think queries perform very purely with no stats. If users convey okay and no issues we can confidently disable it.

With the auto-analyze feature my thought is to offer enhanced automation experience and eliminate the need for each user to perform something via cron job and reduce the pain. But don't wish to take away any of existing functionality or cause disruption in current behaviour.

File: src/backend/catalog/namespace.c (hunk 2 above)
Comment (ashwinstar) [reply src/backend/catalog/namespace.c:3306]: Something on below lines would be helpful....

FIXME: This function doesn't work or useful in GPDB (only useful for utility mode temp tables which are none or rare). Since the temp namespace for QD and QE is using gp_session_id as suffix instead of backendID. Function needs to be modified to work for GPDB. Maybe checking if `gp_session_id` is active in system or not currently. Only user of this function is autovacuum process so far so the impact is low.

File: gpMgmt/bin/gpinitsystem (hunk 1 above)
Comment (ashwinstar) [reply gpMgmt/bin/gpinitsystem:1214]: Problem is clear to me. I am curious, is there no other way (instead of disabling the guc) to avoid starting autovacuum launcher process on segments if `autovacuum=on` if we don't wish it to launch and only want to have it for template0 xact id wrap around case ?

**Comment by ashwinstar** [comment]
> > > Disable autovacuum for specific tests.
> > 
> > 
> > For these category of tests instead of globally flipping autovacuum GUC, would it be better to create tables in those tests with `autovacuum_enabled = false` instead and keep it simple and faster.
> 
> @ashwinstar For sub-case 2, we could set `autovacuum_enabled = false`. But for sub-case 1, if `autovacuum=on`, autostats will be skipped too. The logic is, we don't need auto-ANALYZE and autostats at the same time since they are doing the same thing. So if `autovacuum=on`, autostats is disabled. And using the auto-ANALYZE logic to execute ANALYZE. Otherwise, things get confusing.
> Since I just group the tests which need to turn off autovacuum, so this does not generate more diffs.
> But if you think `autovacuum_enabled = false` is better, I'll fix it.

Given the number of tests requiring disable autovacuum, `autovacuum_enabled = false` seems very painful option. So, what you have seems fine. Though I dislike spreading `disable_autovacuum` in schedule file.

- Can we do it in some elegant way? Maybe (just a thought) disable autovacuum for full all the tests except autovacuum tests themselves instead selectively this way? Or have a separate section in schedule file for autovacuum disabled tests and only flip it once. 
- How would I know when adding new tests if autovacuum should be disabled or not? What criteria needs to be met? Locally generally I would run single test or even in PR pipeline nothing may get exposed to pose the need for it.
-  I am surprised upstream doesn't face this issue and why so many GPDB tests are so sensitive to it?

**Comment by JunfengYang** [comment]
> * Can we do it in some elegant way? Maybe (just a thought) disable autovacuum for full all the tests except autovacuum tests themselves instead selectively this way? Or have a separate section in schedule file for autovacuum disabled tests and only flip it once.

Our cases are grouped by test sceneries, for example, ao table, vacuum related, and etc. I was afraid to beak these groups.
But If it's ok to not group them together, I think it's better to have a separate section.

> * How would I know when adding new tests if autovacuum should be disabled or not? What criteria needs to be met? Locally generally I would run single test or even in PR pipeline nothing may get exposed to pose the need for it.

Maybe I need write a READEME for this and put it under `src/test`.

> * I am surprised upstream doesn't face this issue and why so many GPDB tests are so sensitive to it?

Upstream actually doesn't have too many test cases and some of the tests are fast enough to run before any `autovacuum` gets triggered. And also their cases add lots of `ANALYZE`, so the `explain` does not get affected.

File: src/test/isolation2/sql/crash_recovery_redundant_dtx.sql [hunk 4]
```diff
@@ -10,7 +10,7 @@
 -- wait till checkpoint reaches intended point
 2:select gp_wait_until_triggered_fault('checkpoint_dtx_info', 1, 1);
 -- the 'COMMIT' record is logically after REDO pointer
-2&:insert into crash_test_redundant values (1);
+2&:insert into crash_test_redundant values (1), (2), (3);
```
[the remark this answers is not in the document]
Comment (JunfengYang) [reply src/test/isolation2/sql/crash_recovery_redundant_dtx.sql:13]: Cause `insert into crash_test_redundant values (1);` using one-phase commit.

When autostats executed, it'll use two-phase commit since the joined segments are more than one for the current transaction.

**Comment by JunfengYang** [comment]
> I think we should explicitly discuss on gpdb-users mailing list if turning off gp_autostats_mode to none is good idea or not.

If auto-ANALYZE enabled, it should use auto-ANALYZE as Postgres does, there no need to turn off `gp_autostats_mode `. And shouldn't make then both work for below reason:

A cluster with `autovacuum=on`, and create a table with `autovacuum_enabled = false`.
As a user, I would expect the table will always have empty stats unless I manually run ANALYZE.
But if we still run autostats with `autovacuum=on`. The table will have stats.
So we should not run autostats with `autovacuum=on`.

And if you mean whether to enable auto-ANALYZE as default, we can continue the discussion in https://groups.google.com/a/greenplum.org/g/gpdb-dev/c/bAyw2KBP6yE.
Then we can evaluate more pros and cons to this.

File: gpMgmt/bin/gpinitsystem (hunk 1 above)
Comment (JunfengYang) [reply gpMgmt/bin/gpinitsystem:1214]: I suppose yes. Base on the description in guc.c.
```
{"autovacuum", PGC_SIGHUP, AUTOVACUUM,
			gettext_noop("Starts the autovacuum subprocess."),
			NULL
		},
```

**Comment by ashwinstar** [comment]
@JunfengYang I pushed few commits to this PR. Mainly to showcase not using gpinitsystem to enable `autovacuum`. Feel free to take them and build on top or modify as see best. I felt this was easiest way instead of writing comments to explain what I was proposing. Also, I locally ran ICW and 2 tests failed with these changes (didn't look into details) but overall seems doesn't break much :-)

**Comment by JunfengYang** [comment]
@ashwinstar Realy appreciate! I'm busy working on other stories currently. I'll take over your idea and fix all failures soon. Thanks for pushing this feature!

**Comment by JunfengYang** [comment]
@ashwinstar @zhangh43 The PR is ready now. Please help me check it again. Thanks!

File: src/test/isolation2/output/autovacuum-analyze.source [hunk 5]
```diff
@@ -368,6 +368,128 @@ select relpages, reltuples from pg_class where oid = 'anaabort'::regclass;
...
+ relpages | reltuples 
+----------+-----------
+ 6        | 3000      
+(1 row)
+
+-- Check analyze count.
+select analyze_count, autoanalyze_count from pg_stat_all_tables_internal where relname = 'autostatstbl';
+ analyze_count | autoanalyze_count 
+---------------+-------------------
+ 2             | 1
```
Comment (ashwinstar) [inline src/test/isolation2/output/autovacuum-analyze.source:490]: The test is little hard to follow along, so adding comment on exact scenario will be good at start of test.

I am still trying to figure out based on this how can we confirm when auto-stats kicks-in auto-analyze didn't kick in. Maybe we need to print `n_mod_since_analyze` after auto-stats kicks-in to make sure, `n_mod_since_analyze` value is reset and hence auto-analyze will not kick-in?

File: src/test/regress/sql/disable_autovacuum.sql [hunk 6]
```diff
@@ -0,0 +1,4 @@
+-- start_ignore
```
Comment (ashwinstar) [inline src/test/regress/sql/disable_autovacuum.sql:1]: why is this in ignore blocks?

File: src/test/regress/greenplum_schedule [hunk 7]
```diff
@@ -19,7 +19,11 @@
 # might not happen if other backends are holding transactions open, preventing
 # vacuum from removing dead tuples. And run gp_toolkit early to make some log
 # file related tests (like gp_log_system, etc) faster.
+test: disable_autovacuum
 test: gp_toolkit
+# dispatch should always run seperately from other cases.
+test: dispatch
```
Comment (ashwinstar) [inline src/test/regress/greenplum_schedule:25]: why dispatch tests need auto-analyze disabled?

File: src/test/isolation2/output/uao/compaction_full_stats.source [hunk 8]
```diff
@@ -41,13 +41,11 @@ SELECT relname, reltuples FROM pg_class WHERE relname = 'foo_index';
 -----------+-----------
  foo_index | 85        
 (1 row)
-SELECT segno, tupcount,state FROM gp_ao_or_aocs_seg('foo');
+-- in case there's autovacuum worker running in the backend, the aoseg will not be dropped which has state = 2
```
Comment (ashwinstar) [inline src/test/isolation2/output/uao/compaction_full_stats.source:44]: Doesn't `VACUUM FULL` hold exclusive lock, so this case shouldn't arise?

File: src/test/isolation2/output/uao/compaction_utility_insert.source [hunk 9]
```diff
@@ -22,10 +22,8 @@ UPDATE foo SET b = -1 WHERE a = 3;
 UPDATE 1
 VACUUM foo;
 VACUUM
-SELECT segment_id, segno, tupcount, state FROM gp_ao_or_aocs_seg('foo');
+SELECT segment_id, segno, tupcount, state FROM gp_ao_or_aocs_seg('foo') where state = 1 and tupcount > 0;
```
Comment (ashwinstar) [inline src/test/isolation2/output/uao/compaction_utility_insert.source:25]: No fault of this PR, though I don't know what this test is testing? Seems before and after vacuum this state will be same

File: src/test/isolation2/output/uao/vacuum_self_serializable.source [hunk 10]
```diff
@@ -25,16 +25,14 @@ SELECT *, segno, tupcount FROM gp_ao_or_aocs_seg('ao');
 (3 rows)
 VACUUM ao;
 VACUUM
-SELECT *, segno, tupcount FROM gp_ao_or_aocs_seg('ao');
+-- in case there's autovacuum worker running in the backend, the aoseg will not be dropped which has state = 2
```
Comment (ashwinstar) [inline src/test/isolation2/output/uao/vacuum_self_serializable.source:28]: Well I don't really like this kind of filtered query. As it just hides bugs if any in state/modcount change happens due to bugs in code. I am okay if wish to disable autovacuum for such tests instead of modifying the query or disable auto-vacuum for the tables created in such tests.

**Review by ashwinstar — COMMENTED** [review]
High-level the code in PR looks in much better shape to me. We will have to squash and rearrange commits though.

Plus will need to modify text in some commit messages as not relevant anymore like
- Enable the auto-ANALYZE for the cluster as default through gpinitsystem
- Test cases related to autostats. These tests are testing the                                                                              
  autostats, so we have to disable autovacuum.

File: src/backend/postmaster/README.auto-ANALYZE [hunk 11]
```diff
@@ -1,50 +1,42 @@
...
+Currently auto-ANALYZE is not supported for partition root/mid-layer tables,
+because we don't have accurate pgstat for mid-layer tables and root tables.
+This is same with Postres.
+There's already a discussion on Postgres upstream to enable auto-ANALYZE for
+root/mid-layer partition tables. Once it merged into upstream, we can cherry
+pick the patch. See:
+https://www.postgresql.org/message-id/CAKkQ508_PwVgwJyBY=0Lmkz90j8CmWNPUxgHvCUwGhMrouz6UA@mail.gmail.com
+
+For relation who's storage parameter `autovacuum_enable=false` with autovacuum enabled,
+neither autovacuum ANALYZE nor `auto_stat`s will execute for it. Users need to manually
```
Comment (zhangh43) [inline src/backend/postmaster/README.auto-ANALYZE:37]: "neither autovacuum ANALYZE nor `auto_stat`s will execute for it" is it still true now? They are controlled by separate GUC, right?

File: src/test/isolation2/output/autovacuum-analyze.source (hunk 5 above)
Comment (zhangh43) [reply src/test/isolation2/output/autovacuum-analyze.source:490]: Another suggestion is to test it's OK to have auto-analyze and auto-stats running at the same time. For example we have concurrent inserts. Some big inserts will trigger auto-stats, while other small inserts will trigger auto-analyze.

File: src/test/regress/sql/disable_autovacuum.sql (hunk 6 above)
Comment (JunfengYang) [reply src/test/regress/sql/disable_autovacuum.sql:1]: Since the output doesn't matter.

File: src/test/regress/greenplum_schedule (hunk 7 above)
Comment (JunfengYang) [reply src/test/regress/greenplum_schedule:25]: This is from the original comment.
It'll check dispatch. And auto-ANALYZE will have its own dispatch.

File: src/test/isolation2/output/uao/compaction_full_stats.source (hunk 8 above)
Comment (JunfengYang) [reply src/test/isolation2/output/uao/compaction_full_stats.source:44]: The backend running transaction may access the table after the `VACUUM FULL` transaction, so the data can not be dropped. Just like the recently dead tuples for heap table.
And autovacuum worker may start a transaction before the `VACUUM FULL` transaction.

File: src/backend/postmaster/README.auto-ANALYZE (hunk 11 above)
Comment (JunfengYang) [reply src/backend/postmaster/README.auto-ANALYZE:37]: Will correct it.

File: src/test/isolation2/output/uao/vacuum_self_serializable.source (hunk 10 above)
Comment (JunfengYang) [reply src/test/isolation2/output/uao/vacuum_self_serializable.source:28]: For this kind of test, actually, the original query will also hide bugs since it only tests the `vacuum` without backend transaction.

File: src/test/isolation2/output/autovacuum-analyze.source (hunk 5 above)
Comment (JunfengYang) [reply src/test/isolation2/output/autovacuum-analyze.source:490]: Add more comments and show `n_mod_since_analyze` when ANALYZE gets executed.

Comment (JunfengYang) [reply src/test/isolation2/output/autovacuum-analyze.source:490]: @ashwinstar 
There's a pgstat issue when executing ANALYZE with insert in the same transaction.
See the below example (test on both GPDB and Postgres master branch):

```
demo=# CREATE TABLE autostatstbl (id bigint);
CREATE TABLE
demo=# begin;
BEGIN
demo=*# INSERT INTO autostatstbl select i from generate_series(1, 1000) as i;
INSERT 0 1000
demo=*# analyze autostatstbl;
ANALYZE
demo=*# end;
COMMIT
demo=# select analyze_count, autoanalyze_count, n_mod_since_analyze from pg_stat_all_tables where relname = 'autostatstbl';
 analyze_count | autoanalyze_count | n_mod_since_analyze
---------------+-------------------+---------------------
             1 |                 0 |                1000
(1 row)
```
You can see that the ANALYZE doesn't clean the `n_mod_since_analyze`.
This is because the ANALYZE call `pgstat_report_analyze` immediately
https://github.com/greenplum-db/gpdb/blob/cb2f8c0337640e918b646df25dc98a284531a7be/src/backend/commands/analyze.c#L969-L978

but the INSERT statement's pgstat change only gets sent out(`pgstat_report_stat`) after commit and ready for the next query.
https://github.com/greenplum-db/gpdb/blob/cb2f8c0337640e918b646df25dc98a284531a7be/src/backend/tcop/postgres.c#L5100-L5140

Since `auto_stats` always execute in the same transaction for the statement that modifies tuples, so it'll not prevent auto-ANALYZE get triggered.
That means even `auto_stats` ANALYZEd the table, auto-ANALYZE could be triggered with the same change. It's not like what we thought at the beginning.

I'll modify test case to cover it and maybe we need to document this behavior.

Comment (ashwinstar) [reply src/test/isolation2/output/autovacuum-analyze.source]: That's annoying. Sure, yes lets document the same and move forward. Ideally, would be good in long term if upstream fixes this issue (so its something we can continue to explore outside of this PR).

File: src/test/regress/sql/disable_autovacuum.sql (hunk 6 above)
Comment (ashwinstar) [reply src/test/regress/sql/disable_autovacuum.sql:1]: why output doesn't matter? if the alter or reload command fails we need to know about it. Output should be ignored in tests only if there is genuine reason to ignore it, I don't see any reason here.

Comment (JunfengYang) [reply src/test/regress/sql/disable_autovacuum.sql:1]: Sure, will fix it.

File: src/test/isolation2/output/uao/vacuum_self_serializable.source (hunk 10 above)
Comment (JunfengYang) [reply src/test/isolation2/output/uao/vacuum_self_serializable.source:28]: Let's create a story to have a more reasonable test for this.

**Comment by JunfengYang** [comment]
@ashwinstar @zhangh43  Is it ready for approval?

**Review by ashwinstar — APPROVED** [review]
Fix typo (auytovacuum -> autovacuum) in message for commit "Enable auto-ANALYZE as defalut for the cluster"

File: src/test/README.autovacuum [hunk 12]
```diff
@@ -0,0 +1,39 @@
+Since we enable auto-ANALYZE as defaule, so when adding new test, please
```
Comment (ashwinstar) [inline src/test/README.autovacuum:1]: s/defaule/default

File: src/test/README.autovacuum [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+consider whether the case will be affected by the autovacuum worker process,
+it may become flaky.
+
+There are 3 classes of cases that need to deal with autovacuum auto-ANALYZE
```
Comment (ashwinstar) [inline src/test/README.autovacuum:5]: Frankly, this README is not useful for future use at all. I suggest not including it. We needed something which helped our future selves to understand the impact of autovacuum and when need to disable it. The README in current form discusses the past and not really the future. So, feel free to exclude it.

**Review by ashwinstar — APPROVED** [review]

File: src/test/README.autovacuum (hunk 13 above)
Comment (JunfengYang) [reply src/test/README.autovacuum:5]: Sure, I was trying to explain the autovacuum may cause the test case flaky, and why the flaky happens.
If this is not needed, I can remove it.

File: src/backend/postmaster/pgstat.c [hunk 14]
```diff
@@ -4565,68 +4565,71 @@ pgstat_send_bgwriter(void)
...
-			pq_beginmessage(&buf, 'y');
-			pq_sendstring(&buf, "PGSTAT");
-			initStringInfo(&stat_data);
-		}
+		/*
+		 * No need to send catalog table's pgstat to QD since if the catalog
+		 * table get updated on QE, QD should have the same update.
+		 */
+		if (tabstat->t_id < FirstNormalObjectId)
+			continue;
```
Comment (hidva) [inline src/backend/postmaster/pgstat.c:4593]: if `tabstat->t_id` is distributed replicated, there should be only one segment that can send the stats.

Comment (JunfengYang) [reply src/backend/postmaster/pgstat.c:4593]: It's hard to decide which segment should send the stats. Actually, I plan to only accept stats from one segment on QD.
But it should be another PR since the current one is really huge.

Comment (hidva) [reply src/backend/postmaster/pgstat.c:4593]: Only content 0 can send the tabstats of replicated tables?

Comment (JunfengYang) [reply src/backend/postmaster/pgstat.c:4593]: I'm afraid hard code specifications may not get approval ^^!

Comment (hidva) [reply src/backend/postmaster/pgstat.c:4593]: Or only the segment whose content is `tableoid%getgpsegmentCount()` could send the stats? This can also get a more balanced load.

Comment (JunfengYang) [reply src/backend/postmaster/pgstat.c:4593]: Yes, we can have a separate PR focus on this part.

**Comment by hidva** [comment]
I think we can introduce a template engine to generate regress tests based on the regress test template. The content of regress test template as follow(based on Liquid):

```liquid
select n_tup_ins, n_tup_upd, n_tup_del, n_tup_hot_upd, n_live_tup, n_dead_tup, n_mod_since_analyze from pg_stat_all_tables where relid = 'table_for_truncate_abort'::regclass;
{% if 'replicated' in amit_conf_distby %}
{% comment %}
ORCA can't be used in the replicated table, the update of table_for_iud will not be implemented by delete + insert
{% endcomment %}
1600|301|777|0|777|1124|777
{% else %}
1901|0|1078|0|777|1124|777
{% endif %}
{% if 'replicated' not in amit_conf_distby %}
{% comment %}
ERROR:  PARTITION BY clause cannot be used with DISTRIBUTED REPLICATED clause
So don't do the following test in DISTRIBUTED REPLICATED
{% endcomment %}
-- Test pgstat table stat for partition table on QD
CREATE TABLE rankpart (id int, rank int, product int)
with ({{amit_conf_opts}})
{{amit_conf_distby}} PARTITION BY RANGE (rank)
( START (1) END (10) EVERY (5),
  DEFAULT PARTITION extra );
...
{% endif %}
```

`amit_conf_distby` will be replaced by `appendonly=false`, `appendonly=true,orientation=row`, `appendonly=true,orientation=column`, and `amit_conf_distby` will be replaced by `distributed randomly`, `distributed replicated`, `''`(an empty string which means `distributed by`). So that we can test pgstat_qd_tabstat on heap/ao_row/ao_column and/or `distributed by`/`distributed replicated`/`distributed randomly`.

```
test pgstat_qd_tabstat-heap_hash_dist ... ok (83.60 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-heap_rand_dist ... ok (83.65 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-heap_replicated ... ok (63.62 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoro_hash_dist ... ok (84.45 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoro_rand_dist ... ok (84.58 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-aoro_replicated ... ok (64.32 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoco_hash_dist ... ok (86.36 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoco_rand_dist ... ok (86.47 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-aoco_replicated ... or (65.70 sec)  (diff:0.25 sec)
```

**Comment by JunfengYang** [comment]
@hidva Fell free to open PR for it. Seems not the right place to discuss under this PR now.

**Comment by JunfengYang** [comment]
Merged, thanks!

**Comment by LittleWuCoding** [comment]
Is there any plan that merge those commits to 6X_STABLE ?

**Comment by hidva** [comment]
> Is there any plan that merge those commits to 6X_STABLE ?

It seems that there is no plan... @JunfengYang

**Comment by JunfengYang** [comment]
Not for now, but we may consider it.

**Comment by ashwinstar** [comment]
> > Is there any plan that merge those commits to 6X_STABLE ?
> 
> It seems that there is no plan... @JunfengYang

There are lot of improvements in master branch as its based on PostgreSQL 12 for vacuum and auto-vacuum areas, which are missing in GPDB6. So, we will need to be really cautious to introduce this feature without those improvements in GPDB6. Hence, I would say possibility of us taking this feature in STABLE GPDB6 is very less.
